### PR TITLE
Add macos_pref module for direct access to Core Foundation Preferences API

### PR DESCRIPTION
**Pull request by kinglouie**
##### SUMMARY
This is a new module that allows users to manipulate macOS preferences using the Core Foundation API directly instead of wrapping defaults or standalone plist tools. By using the operating system's preferences API, osx_pref ensures that changes go through the same caching logic in cfprefsd that is used by native applications. In contrast with methods that wrap the defaults command, using the API allows users to access preferences with complex data structures like nested dictionaries.

To assist with manipulating preferences in nested dictionaries (e.g. in the com.apple.finder domain), in addition to the default replace strategy for writing values, macos_pref also implements a deep merge mechanism so that users only need to specify keys and values that they care to change.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
macos_pref

##### ANSIBLE VERSION
```
ansible 2.6.0
```

##### ADDITIONAL INFORMATION

This was originally only planned as a resubmission of ansible/ansible-modules-extras#2392 since the original author didn't seem to bother resubmitting after the relocation of ansible modules. However as I was working on the code to make it code standard compliant, I discovered that it lacked a lot of useful features which are now included.

### CODE REVIEW

**Comment by kinglouie** [comment]
Sorry for the commit gore, this is my first time using python and originally I only planned to resubmit the code from the original author. At that time I didn't know that the contributing guidelines for new modules changed that much.

**Comment by kinglouie** [comment]
I'm thinking about removing the option `action` and determine the action based on the existence of the `value` option. Any thoughts on this?

**Comment by azaghal** [comment]
Couple of things pop to mind here (mind you, I'm not a Mac user, so can't test or have deeper insight into this stuff - and I'm just a minor contributor ;):

1. Better to have explicit `action` - but default to `set`. That should make it pretty consistent with the rest of modules.

2. Is it possible to remove an option? Is it possible to reset an option to default? In that case it will be much easier to have an explicit `action`. Also, maybe consider adding ability to unset/remove option.

3. Now, I am maybe a bit biased on this (from working on the `dconf` module), but would it make sense to rename `action` to `state` and have possible values as `present`/`absent`(if applicable)/`read`?

4. Perhaps `dict_set_method` could be moved into `action` (or `state` if you like **(3)**)? E.g. have it as `merge` there? This could reduce option clutter perhaps. Any case where one would benefit from having those two separate?

**Comment by kinglouie** [comment]
@azaghal Thank you for your feedback!
I just thought about removing it because it is a little redundant since there are only two actions and one of them requires `value` to be present. 

Addressing you input:

2. It is possible to remove options, at the current state of the module you pass `null` as `value` to remove the key. 

3. I like your suggestions to rename `action` to `state` but I still would not add the `read` action to state because reading an option doesn't affect the state in any way.

4. I also like this, so I would end up with these possible values for `state`: `["absent", "replace", "merge"]` with `replace` as default value.

Here are a few examples with my and your proposed changes:

#### Set key with merge:
```
- name: Configure finder kit standard view
  osx_pref:
    key: FK_StandardViewSettings
    value:
      IconViewSettings:
        iconSize: 36
    state: merge
    domain: com.apple.finder
```
#### Set key with replace (default):
```
- name: Configure finder kit standard view
  osx_pref:
    key: FK_StandardViewSettings
    value:
      IconViewSettings:
        arrangeBy: name
        gridSpacing: 45
        iconSize: 36
        showIconPreview: true
        showItemInfo: false
        labelOnBottom: true
        textSize: 12
    domain: com.apple.finder
```
#### Delete key:
```
- name: Delete Key
  osx_pref:
    key: FK_StandardViewSettings
    state: absent
    domain: com.apple.finder
```
#### Read key:
```
- name: Read finder kit standard view
  osx_pref:
    key: FK_StandardViewSettings
    domain: com.apple.finder
  register: finder_view_style
```

File: lib/ansible/modules/system/osx_pref.py
Comment (pilou-) [inline lib/ansible/modules/system/osx_pref.py]: Use [formatting function](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#formatting-functions) `C(...)` for sample values.

File: lib/ansible/modules/system/osx_pref.py
Comment (pilou-) [inline lib/ansible/modules/system/osx_pref.py]: line seems over indented.

**Review by pilou- — COMMENTED** [review]
- Would not it be useful to always return value ? For example when 'merge' is used.
- i like the proposal (https://github.com/ansible/ansible/pull/35683#issuecomment-364058748)
- tests must be added, do not hesitate asking if any help is needed

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: Any limitations on what is the minimal OS version supported, or should be work with pretty much anything?

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: Any limitations on what version of library must be used, or should it work with pretty much any?

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: Is this some kind of ability to administer multiple machines from same Mac? Maybe mention how this differs from the host Ansible is running on (not sure how common is knowledge about fleet management amongst somewhat regular Mac users, I guess).

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: Would it be safer to default to `currentHost` instead? (provided that `anyHost` means updating settings on multiple machines within network)

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: Document that if value is not provided, the preference value will be read instead. Might be useful to mention this in the main description too.

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: This looks a bit dangerous to me, and does mean a person has to figure out what parameters need to be passed-in into the constructor based on existing calling code. I would recommend listing the possible argument to `__init__` explicitly. Processing `kwargs` in this way should probably be left for cases where the implementation actually deals with variable set of parameters.

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [inline lib/ansible/modules/system/macos_pref.py]: You should add an explicit code snippet below to verify that if state is replace/merge, the value must be provided. Have you tested what happens if state is set to replace/merge, but no value is provided?

**Review by azaghal — CHANGES_REQUESTED** [review]
A couple of minor questions for clarification, some tiny docs improvements could be useful, and a couple small changes to code for better robustness. Otherwise looks solid (although, again, no Mac to test on here :)

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: This is not about multiple machines
The "default" for preferences is `anyHost`. Using the bundle name com.example for an example, these preferences would be stored at

```~/Library/Preferences/com.example.plist```
What makes this "anyHost" is the assumption that your home directory is stored on a network share. There is no additional mechanism to synchronize it across computers. Generally speaking, you should always use anyHost unless you are setting a value which really does need to be specific to the physical machine the user is sitting at.

Preferences which are set for the current host are stored under the ByHost directory, with a file name that includes a UUID for the computer, such as:

```~/Library/Preferences/ByHost/com.example.74263C97-28D5-4AF0-8E7A-3169F1CCB545.plist```

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: the default is anyHost, see https://github.com/ansible/ansible/pull/35683#discussion_r167241839

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: Will update that 👍

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: Per design if `value` is ommited, the module just returns (reads) the value of `key`. See the last example here https://github.com/ansible/ansible/pull/35683#issuecomment-364058748
In that example `state` is also ommited which means it defaults to `replace`.

You can find that logic [here](https://github.com/kinglouie/ansible/blob/0e23532dff2bad7ff4aa7a5429f58db8c5c72a72/lib/ansible/modules/system/macos_pref.py#L271-L279)
 I think it is simple enough to not confuse users.

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: Will update that 👍

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [reply lib/ansible/modules/system/macos_pref.py]: Hm... Yeah, you're right on that one. Although... I can actually now see why having `read` as a state could help reduce some confusion/improve consistency. For example, the following samples will read parameter in spite of having the explicit state that denotes changes to attributes:

```
- name: Show all files
  macos_pref:
    domain: com.apple.finder
    key: AppleShowAllFiles
    state: merge

- name: Show all files
  macos_pref:
    domain: com.apple.finder
    key: AppleShowAllFiles
    state: replace
```

A user that forgets to provide value might end-up thinking the invocations have actually changed the value. So, from usability perspective (at expense of being more expressive for the `read` operation), one could produce more explicit error information for the user.

Your call on this one, though :)

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [reply lib/ansible/modules/system/macos_pref.py]: Understood, sounds good then.

File: lib/ansible/modules/system/macos_pref.py
Comment (azaghal) [reply lib/ansible/modules/system/macos_pref.py]: How about mentioning something along the lines of "Unless you know what you are doing, keep `anysoft` as value" and/or maybe linking to/mentioning some official Apple docs on the topic?

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: I see your concerns but I think a user forgetting to provide a value when he wants to set something is rare enough it can be neglected in favor of fewer options to be set in 99% of the cases.

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: I couldn't find any info what the minimum version of macOS/OS X is where PyObjC is supported, I will investigate that matter.

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: I will investigate that. 👍

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py]: Will update 👍

File: lib/ansible/modules/system/macos_pref.py
Comment (etienned) [reply lib/ansible/modules/system/macos_pref.py]: PyObjC was included in Mac OS X from version 10.5 (Leopard), but python included in 10.5 was version 2.5. Personally, I've developed my `osx_defaults` version #24808 on Mac OS X 10.6 (python 2.6) but test it and use it also on macOS 10.11, 10.12 and 10.13 (python 2.7). But I should add that it's possible to install PyObjC on a system with the system's provided python or another python installation.

Comment (azaghal) [reply lib/ansible/modules/system/macos_pref.py]: Well, I guess that anything that's actually still supported will work. Maybe mentioning minimal requirement as 10.11+?

**Comment by kinglouie** [comment]
ready_for_review

**Review by mattclay — DISMISSED** [review]
Please add the following to the aliases file so that the integration tests can run in CI:

```
posix/ci/group1
skip/freebsd
skip/rhel
destructive
```

File: test/integration/targets/macos_pref/aliases
Comment (mattclay) [inline test/integration/targets/macos_pref/aliases]: Include a newline here.

**Comment by kinglouie** [comment]
ready_for_review

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: This is probably something you want to add to the `notes:` section ?

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: If it's not required, you don't need to add this.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: If it's not required, you don't need to add this.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: If it's not required, you don't need to add this.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: If it's not required, you don't need to add this.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: The default value is null, so you don't have to provide this information.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: If it's not required, you don't need to add this.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: If it's not required, you don't need to add this.
Besides, it has a default value, so in a way it is required ;-)

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py:93]: In the examples you could show an example using Jinja2 to convert to base64 ?

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py:112]: We don't tend to add checks in the examples, but if we do it's probably better to use the 'assert' module.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py:209]: I wonder whether `data` is not a better name than value.
I think it's used more often by other modules.

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: I prefer to see this in collapsed view, rather than expanded.
Also, if a parameter is not required, you don't need to add it explicitly.
(It's easier to see at a glance which parameters are required)

File: lib/ansible/modules/system/macos_pref.py
Comment (dagwieers) [inline lib/ansible/modules/system/macos_pref.py]: Default None values do not have to be added, because that is the default.
It's implicit that unset parameters always return None (and that's one way to check for a value)

**Review by dagwieers — CHANGES_REQUESTED** [review]
Looks good, but I am not able to test the module (lacking macOS devices).

**Comment by dagwieers** [comment]
Wondering if there's an interest in starting a macOS Working Group to look more actively at what modules would be useful for managing macOS devices (or even other Apple devices ?) and act as a steward for macOS users.

If there's an interest I can provide details how to get it started.

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py:209]: macOS preferences are key/value pairs, so shouldn't you get a `value` in return when you read a `key`? When setting a key you provide the value in `value` as well.

**Comment by etienned** [comment]
@dagwieers I would like to point out that this module proposition is mostly the code I did proposed #24808 (with a few changes and added features) as a backward compatible upgrade to the already existing `osx_defaults` module. The kinglouie version is not backward compatible and introduce a new module that have the same function as `osx_defaults`. I think this should be discuss before accepting this PR.

**Comment by dagwieers** [comment]
I would probably raise this in a core meeting by adding an item to the Core meeting agenda at: https://github.com/ansible/community/issues/296

So for new modules, it's important that sufficient people have reviewed your code and approve of it. If you don't feel it is ready, you can prefix the title using `[WIP]`. You can ask on IRC for people to review your module. Once you are confident the module is ready and it has been reviewed/approved, and you can't find someone who's willing to merge it, you can bring it up in a Core meeting to get things moving again.

Once your module has been accepted, merging changes into it is usually quite simple. Since you (or any other author/maintainer in the module docs or in .github/BOTMETA.yml) have the ability to approve PRs to your module. This is explained in the ansibot documentation: https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md

Asking for advice in #ansible-devel is always a good way to start if you have any questions.

**Comment by dagwieers** [comment]
BTW My suggestion to create a macOS related Working Group could help to maintain a set of modules in group as there's no longer just a single maintainer per module. You can assign maintainership to a group of people and discuss/decide about future changes and needed improvements.

I explained the process a bit to @kinglouie but let me know if you're interested or need help setting up a Working Group. The main goal is to make it more practical to collaborate on modules in a consensus model and think about the long-term process. Rather than e.g. dropping a single module in the repository and never look back.

And in this case to avoid duplicate effort and smoother/more efficient contributions...

File: lib/ansible/modules/system/macos_pref.py
Comment (kinglouie) [reply lib/ansible/modules/system/macos_pref.py:93]: Somethink like this?
```
# Save an image to a data key
- name: Set user avatar
  macos_pref:
    domain: com.ansible.example
    key: UserAvatar
    value: "{{ lookup('file', '/etc/avatar.jpg') | b64encode }}"
```

**Comment by etienned** [comment]
@dagwieers Thanks for all your advices. There's a lot of things necessary to read/understand to contribute to Ansible. But I think the one I need to remember the most is asking advice on IRC!

Just to be clear my PR is not a new module but a rewrite of an existing module.

About the macOS related Working Group, I don't know if it's a necessity? There's only a few modules proper to macOS: homebrew, homebrew_tap, homebrew_cask, osx_defaults, osx_say, if I'm not wrong.

**Comment by dagwieers** [comment]
@etienned A macOS-related Working Group could be looking at current gaps, improving macOS-specific use-cases, specific macOS integration, and collaborate as a group. So it would be more than just the handful of existing modules.

I am not a macOS user myself, but what about patch management, (kernel) drivers, maybe keychain/vault integration, improvements to macOS-related facts, iTunes software installation, system reboots, elevated rights other than sudo ?, ... I remember companies running MacOSX on servers, not sure if that is still a thing ?

**Comment by etienned** [comment]
@dagwieers I understand what you means. I just don't know if there's enough demands for those needs (patch management, (kernel) drivers, maybe keychain/vault integration, etc.) that will justify such a group?  And how to find if there's an interest?

For your information Apple do not sell server anymore but they have [macOS server](https://www.apple.com/ca/macos/server/) targeted to small businesses.

Unfortunately, as you could had deduced from my answering delay, I'm not currently the man that will lead this. I'm already wearing too much hats.

**Comment by kinglouie** [comment]
@etienned Are you still interested in getting this/your module into ansible core? I waited some time to see if you're following through to get this module merged but it seems there is nothing happening. I don't want to put in the effort to push this PR if you're not okay about it but I also want to get this module or your module into ansible pretty badly because I wan't to update my role [kinglouie/ansible-role-macos](https://github.com/kinglouie/ansible-role-macos) to use the pyobjc bridge.

**Comment by kinglouie** [comment]
@etienned Are you still interested in getting this/your module into ansible core? I waited some time to see if you're following through to get this module merged but it seems there is nothing happening. I don't want to put in the effort to push this PR if you're not okay about it but I also want to get this module or your module into ansible pretty badly because I wan't to update my role [kinglouie/ansible-role-macos](https://github.com/kinglouie/ansible-role-macos) to use the pyobjc bridge.

**Comment by etienned** [comment]
@kinglouie Yes I'm definitely still interested in getting my module into ansible. Unfortunately I've currently no free time to push it. I wish I could find some time in August.

**Comment by dagwieers** [comment]
@kinglouie @etienned Does this module still have relevance ? Can we close this ?

**Comment by kinglouie** [comment]
@dagwieers Although this has a few features which are missing in the other PR, I think it can be closed and I will try to add these later on, when the original PR from @etienned is merged.

**Comment by dagwieers** [comment]
@kinglouie Thanks ! In the meantime we also started a macOS Working Group to collaborate on macOS-related integration. You can find the details at: https://github.com/ansible/community/wiki/macOS

**Comment by dagwieers** [comment]
Closing this in favor of #24808, let's hope we can get this into Ansible v2.8. The deadline is early March, so there's not that much time to get things tested and approved.

**Pull request closed without merging**
